File: telegram-monitor.js
import 'dotenv/config';
import { EmbedBuilder } from 'discord.js';
import { TelegramClient } from 'telegram';
import { StringSession } from 'telegram/sessions/index.js';
import input from 'input';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { activeWars, warKey } from './war-store.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const sessionFile = path.join(__dirname, 'telegram_session.txt');

const processedMessages = new Set();

export async function startTelegramMonitor(discordClient) {
  const apiId = parseInt(process.env.TELEGRAM_API_ID);
  const apiHash = process.env.TELEGRAM_API_HASH;

  let sessionString = '';
  if (fs.existsSync(sessionFile)) {
    sessionString = fs.readFileSync(sessionFile, 'utf-8').trim();
  }

  const session = new StringSession(sessionString);
  const tgOptions = {
    connectionRetries: 5,
    autoReconnect: true,
  };
  if (process.env.TG_PROXY_IP) {
    tgOptions.proxy = {
      ip: process.env.TG_PROXY_IP,
      port: parseInt(process.env.TG_PROXY_PORT),
      socksType: 5,
    };
  }
  const client = new TelegramClient(session, apiId, apiHash, tgOptions);

  await client.start({
    phoneNumber: async () => {
      console.log('\n📱 Введи номер телефона (например +79998887766):');
      return await input.text('Номер: ');
    },
    password: async () => {
      console.log('🔐 Пароль 2FA (если нет — Enter):');
      return await input.text('Пароль: ');
    },
    phoneCode: async () => {
      console.log('📬 Код из Telegram:');
      return await input.text('Код: ');
    },
    onError: (err) => console.error('[TG error]', err),
  });

  fs.writeFileSync(sessionFile, client.session.save(), 'utf-8');
  console.log('✅ Telegram подключен');

  client.addEventHandler(async (event) => {
    try {
      await handleMessage(event, discordClient);
    } catch (e) {
      console.error('[TG handle error]', e.message);
    }
  });
}

async function handleMessage(event, discordClient) {
  const msg = event.message;
  if (!msg || !msg.message) return;

  const text = msg.message;

  // Фильтр: только уведомления о войне нашей организации
  if (!text.includes('Организация:') || !text.toLowerCase().includes('забил')) return;

  const msgId = String(msg.id);
  if (processedMessages.has(msgId)) return;
  processedMessages.add(msgId);

  // Парсинг: "Организация: события | Lux_Aeterna, сервер Rockford"
  const orgMatch = text.match(/Организация:.*?\|\s*(.+?),\s*сервер\s+(\S+)/i);
  // Парсинг: "забила REINHARD войну за Ломбард Strawberry на 21:48, 8х8"
  const warMatch = text.match(/забил[аи]?\s+(.+?)\s+войну за\s+(.+?)\s+на\s+(\d{1,2}:\d{2})/i);

  if (!orgMatch || !warMatch) {
    console.log('[TG] Сообщение не распознано:', text.slice(0, 80));
    return;
  }

  const family = orgMatch[1].trim();
  const server = orgMatch[2].trim();
  const opponent = warMatch[1].trim();
  const point = warMatch[2].trim();
  const time = warMatch[3].trim();

  // Формат боя: "8х8"
  const formatMatch = text.match(/(\d+х\d+|\d+x\d+)/i);
  const format = formatMatch ? formatMatch[1] : '—';

  const side = 'АТАКА'; // "забили войну" = мы атакуем
  const sideEmoji = '⚔️';
  const sideColor = 0xED4245;

  const channel = await discordClient.channels.fetch(process.env.VZP_CHANNEL_ID);
  const guild = channel.guild;
  await guild.roles.fetch();
  const vzpRole = guild.roles.cache.find(r => r.name.includes('| VZP'));
  const mention = vzpRole ? `<@&${vzpRole.id}>` : '';

  const embed = new EmbedBuilder()
    .setTitle(`${sideEmoji} ВЗП — ${side}`)
    .setColor(sideColor)
    .addFields(
      { name: '🆚 Против', value: opponent, inline: true },
      { name: '🗺️ Объект', value: point, inline: true },
      { name: '🖥️ Сервер', value: server, inline: true },
      { name: '⏰ Время', value: time, inline: true },
      { name: '⚔️ Формат', value: format, inline: true },
    )
    .setFooter({ text: '.aeterna • VZP Monitor' })
    .setTimestamp();

  const sent = await channel.send({ content: mention, embeds: [embed] });
  await sent.react('➕');

  // Сохраняем для дальнейшего обогащения с сайта
  const key = warKey(point, server);
  activeWars.set(key, {
    messageId: sent.id,
    channelId: channel.id,
    opponent,
    point,
    server,
    time,
    format,
    side,
    mapAdded: false,
    finished: false,
    createdAt: Date.now(),
  });

  console.log(`[TG] ВЗП создан: ${side} vs ${opponent} | ${point} | ${server} | ${time}`);
}

File: telegram-monitor.py
import os
import sys, io, asyncio, json, os, re
from http.client import HTTPConnection

# Под pythonw.exe (запуск без консоли) sys.stdout/stderr == None.
# Обращение к .buffer крашит скрипт сразу же, поэтому пишем в лог-файл.
if sys.stdout is None or sys.stderr is None:
    _log = open(os.path.join(os.path.dirname(__file__), 'tg-monitor.log'), 'a', encoding='utf-8')
    sys.stdout = _log
    sys.stderr = _log
else:
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')

from telethon import TelegramClient, events
from telethon.sessions import StringSession
from telethon.network.connection import ConnectionTcpMTProxyRandomizedIntermediate

SESSION_FILE = os.path.join(os.path.dirname(__file__), 'telethon_session.txt')
SESSION = open(SESSION_FILE, encoding='utf-8').read().strip()
API_ID = int(os.environ.get("TG_API_ID", "0"))
API_HASH = os.environ.get("TG_API_HASH", "")
PROXY = ('127.0.0.1', 8443, os.environ.get("TG_PROXY_SECRET",""))
NODE_PORT = 3001

def send_to_node(data):
    try:
        payload = json.dumps(data, ensure_ascii=False).encode('utf-8')
        conn = HTTPConnection('127.0.0.1', NODE_PORT, timeout=5)
        conn.request('POST', '/tg-event', payload, {'Content-Type': 'application/json'})
        resp = conn.getresponse()
        print(f'Sent to node: {resp.status}', flush=True)
        conn.close()
    except Exception as e:
        print(f'Node send error: {e}', flush=True)

def parse_war_message(text):
    # "Организация: события | Lux_Aeterna, сервер Rockford"
    # АТАКА:  "Ваша организация забила REINHARD войну за Ломбард Strawberry на 21:48, 8х8"
    # ДЕФ:    "Wright забили Вашей организации войну за Ломбард Strawberry на 22:43, 8х8"
    org_match = re.search(r'Организация:.*?\|\s*\**(.+?)\**,\s*сервер\s+(\S+)', text, re.I)
    if not org_match:
        return None

    # Сначала проверяем деф (на нас напали)
    war_match = re.search(r'([^\n]+?)\s+забил[аи]?\s+Ваш\w*\s+организаци\w*\s+войну за\s+(.+?)\s+на\s+(\d{1,2}:\d{2})', text, re.I)
    side = 'ДЕФ'
    if not war_match:
        # Иначе атака (мы напали)
        war_match = re.search(r'Ваш\w*\s+организаци\w*\s+забил[аи]?\s+(.+?)\s+войну за\s+(.+?)\s+на\s+(\d{1,2}:\d{2})', text, re.I)
        side = 'АТАКА'
    if not war_match:
        return None

    return {
        'type': 'war',
        'family': org_match.group(1).strip(),
        'server': org_match.group(2).strip(),
        'opponent': war_match.group(1).strip(' *'),
        'point': war_match.group(2).strip(),
        'time': war_match.group(3).strip(),
        'side': side,
        'format': (re.search(r'(\d+[хx]\d+)', text) or type('', (), {'group': lambda s, n: '—'})()).group(1),
        'text': text,
    }

async def main():
    client = TelegramClient(
        StringSession(SESSION), API_ID, API_HASH,
        connection=ConnectionTcpMTProxyRandomizedIntermediate,
        proxy=PROXY
    )
    await client.connect()
    me = await client.get_me()
    print(f'TG connected as: {me.first_name}', flush=True)

    @client.on(events.NewMessage(from_users='gta5rp_helperbot'))
    async def handler(event):
        text = event.message.text or ''
        print(f'[TG] New message: {text[:80]}', flush=True)
        if 'Организация:' not in text:
            return
        parsed = parse_war_message(text)
        if parsed:
            send_to_node(parsed)
            print(f'[TG] Sent war event: {parsed["opponent"]} at {parsed["point"]}', flush=True)

    print('TG monitor started, waiting for messages...', flush=True)
    await client.run_until_disconnected()

if __name__ == '__main__':
    asyncio.run(main())

